### ISSUE TITLE
Properly scope jetty-security optional / test dependencies

### ISSUE THREAD

**Issue by uschindler** [eclipse/jetty.project#11024]
**Jetty version(s)**
12.0.4

**Jetty Environment**
core, ee8 (but seems to affect all)

**Java version/vendor** `(use: java -version)`
17

**Description**
jetty-security module has optional 3rd party dependencies included, but not marked as optional.

**How to reproduce?**
Upgrade Jetty to 12.0.4 from 12.0.3. When you resolve the dependencies, jetty-security (one of the defaults that is always included as soon as you use servlets) adds some new 3rd party dependencies (apache directory server API) and many additional 3rd party dependencies transitively, e.g. commons-lang3 which we want to avoid in our infrastructure.

This behaviour is not acceptable, because most people won't use any of those features. It should also not happen in a minor release.

**Possible Solutions**
- Mark those dependencies as optional
- Refactor the directory API code to a separate jetty module

**Comment by uschindler** [comment]
This is caused by #10857.

This is the relevant part of the dependency tree of including `jetty-ee8-servlet` dependency (generated with Apache IVY):
```
[ivy:dependencytree] +- org.eclipse.jetty.ee8#jetty-ee8-servlet;12.0.4
[more other jetty-deps]
[ivy:dependencytree] |  \- org.eclipse.jetty.ee8#jetty-ee8-security;12.0.4
[ivy:dependencytree] |  |  \- org.eclipse.jetty.ee8#jetty-ee8-nested;12.0.4
[ivy:dependencytree] |  |  |  +- org.eclipse.jetty#jetty-http;12.0.4
[ivy:dependencytree] |  |  |  |  +- org.eclipse.jetty#jetty-io;12.0.4
[ivy:dependencytree] |  |  |  |     \- org.eclipse.jetty#jetty-util;12.0.4
[ivy:dependencytree] |  |  |     \- org.eclipse.jetty#jetty-util;12.0.4
[ivy:dependencytree] |  |  |  +- org.eclipse.jetty#jetty-security;12.0.4
[ivy:dependencytree] |  |  |  |  +- org.eclipse.jetty#jetty-server;12.0.4
[ivy:dependencytree] |  |  |  |  |  +- org.eclipse.jetty#jetty-http;12.0.4
[ivy:dependencytree] |  |  |  |  |  |  +- org.eclipse.jetty#jetty-io;12.0.4
[ivy:dependencytree] |  |  |  |  |  |     \- org.eclipse.jetty#jetty-util;12.0.4
[ivy:dependencytree] |  |  |  |  |     \- org.eclipse.jetty#jetty-util;12.0.4
[ivy:dependencytree] |  |  |  |  |  \- org.eclipse.jetty#jetty-io;12.0.4
[ivy:dependencytree] |  |  |  |  |     \- org.eclipse.jetty#jetty-util;12.0.4
[ivy:dependencytree] |  |  |  |  +- org.apache.directory.api#api-asn1-api;2.1.5
[ivy:dependencytree] |  |  |  |     \- org.apache.directory.api#api-i18n;2.1.5
[ivy:dependencytree] |  |  |  |  +- org.apache.directory.api#api-ldap-model;2.1.5
[ivy:dependencytree] |  |  |  |  |  +- org.apache.directory.api#api-asn1-api;2.1.5
[ivy:dependencytree] |  |  |  |  |     \- org.apache.directory.api#api-i18n;2.1.5
[ivy:dependencytree] |  |  |  |  |  +- org.apache.directory.api#api-util;2.1.5
[ivy:dependencytree] |  |  |  |  |     \- org.apache.directory.api#api-i18n;2.1.5
[ivy:dependencytree] |  |  |  |  |  +- org.apache.directory.api#api-i18n;2.1.5
[ivy:dependencytree] |  |  |  |  |  +- org.apache.directory.api#api-asn1-ber;2.1.5
[ivy:dependencytree] |  |  |  |  |  |  +- org.apache.directory.api#api-asn1-api;2.1.5
[ivy:dependencytree] |  |  |  |  |  |     \- org.apache.directory.api#api-i18n;2.1.5
[ivy:dependencytree] |  |  |  |  |  |  +- org.apache.directory.api#api-util;2.1.5
[ivy:dependencytree] |  |  |  |  |  |     \- org.apache.directory.api#api-i18n;2.1.5
[ivy:dependencytree] |  |  |  |  |     \- org.apache.directory.api#api-i18n;2.1.5
[ivy:dependencytree] |  |  |  |  |  +- org.apache.mina#mina-core;2.2.3
[ivy:dependencytree] |  |  |  |  |  +- org.apache.servicemix.bundles#org.apache.servicemix.bundles.antlr;2.7.7_5
[ivy:dependencytree] |  |  |  |  |  +- org.apache.commons#commons-lang3;3.13.0
[ivy:dependencytree] |  |  |  |  |  +- org.apache.commons#commons-collections4;4.4
[ivy:dependencytree] |  |  |  |     \- com.github.ben-manes.caffeine#caffeine;2.9.3
[ivy:dependencytree] |  |  |  |  \- org.apache.directory.api#api-util;2.1.5
[ivy:dependencytree] |  |  |  |     \- org.apache.directory.api#api-i18n;2.1.5
[...]
```

**Comment by sbordet** [comment]
Those should have been test dependencies, sorry about that.
We plan a 12.0.5 release before Christmas.

**Comment by uschindler** [comment]
Thanks!

I added a workaround by excluding all directory deps with Ivy's `<exclude org="org.apache.directory.api"/>`. It should be the same possible for Maven or Gradle.

**Comment by olamy** [comment]
holly crap :(

**Comment by olamy** [comment]
PR https://github.com/jetty/jetty.project/pull/11025

**Comment by joakime** [comment]
Merged PR #11025
